Source/Events.Store.MongoDB/Processing/Streams/StreamProcessorStates.cs

```diff
@@ -27,60 +29,125 @@ public class StreamProcessorStates : IStreamProcessorStateRepository
+        /// Handles <see cref="Partitioned.StreamProcessorState"/> separately also.
+        /// IsUpsert option creates the document if one isn't found.
+        /// </summary>
+        /// <param name="id">The <see cref="StreamProcessorId" />.</param>
+        /// <param name="baseStreamProcessorState">The <see cref="IStreamProcessorState" />.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken" />.</param>
+        /// <returns>A <see cref="Task" /> representing the asynchronous operation.</returns>
+        public async Task Persist(IStreamProcessorId id, IStreamProcessorState baseStreamProcessorState, CancellationToken cancellationToken)
         {
             try
             {
-                return Task.CompletedTask;
-
-                // TODO: There are now two different collections for stream proecssor states. One for StreamProcessorId (Events.Processing) and one for SubscriptionId (EventHorizon)
-                // var states = await _connection.GetStreamProcessorStateCollection(streamProcessorId.ScopeId, cancellationToken).ConfigureAwait(false);
-                // var state = await states.Find(
-                //     _streamProcessorFilter.Eq(_ => _.ScopeId, streamProcessorId.ScopeId.Value)
-                //         & _streamProcessorFilter.Eq(_ => _.EventProcessorId, streamProcessorId.EventProcessorId.Value)
-                //         & _streamProcessorFilter.Eq(_ => _.SourceStreamId, streamProcessorId.SourceStreamId.Value))
-                //     .FirstOrDefaultAsync(cancellationToken)
-                //     .ConfigureAwait(false);
+                if (id is SubscriptionId subscriptionId)
+                {
+                    if (baseStreamProcessorState is StreamProcessorState streamProcessorState)
+                    {
+                        var replacementState = new SubscriptionState(
+                            subscriptionId.ConsumerTenantId,
+                            subscriptionId.ProducerMicroserviceId,
+                            subscriptionId.ProducerTenantId,
+                            subscriptionId.ScopeId,
+                            subscriptionId.StreamId,
+                            subscriptionId.PartitionId,
+                            streamProcessorState.Position,
+                            streamProcessorState.RetryTime,
+                            streamProcessorState.FailureReason,
+                            streamProcessorState.ProcessingAttempts);
+                        var states = await _connection.GetSubscriptionStateCollection(replacementState.ScopeId, cancellationToken).ConfigureAwait(false);
+                        var persistedState = await states.ReplaceOneAsync(
+                            CreateFilter(subscriptionId),
+                            replacementState,
+                            new ReplaceOptions { IsUpsert = true })
+                            .ConfigureAwait(false);
+                    }
```

Comment:
It should throw an exception if stream processor state is not StreamProcessorState